flink-end-to-end-tests/flink-stream-state-ttl-test/src/main/java/org/apache/flink/streaming/tests/verify/AbstractTtlStateVerifier.java

```diff
@@ -0,0 +1,132 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.tests.verify;
+
+import org.apache.flink.api.common.state.State;
+import org.apache.flink.api.common.state.StateDescriptor;
+import org.apache.flink.api.common.state.StateTtlConfiguration;
+import org.apache.flink.api.common.time.Time;
+import org.apache.flink.api.common.typeutils.TypeSerializer;
+import org.apache.flink.runtime.state.FunctionInitializationContext;
+
+import javax.annotation.Nonnull;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Random;
+import java.util.stream.IntStream;
+
+abstract class AbstractTtlStateVerifier<D extends StateDescriptor<S, SV>, S extends State, SV, UV, GV>
+	implements TtlStateVerifier<UV, GV> {
+	static final Random RANDOM = new Random();
+
+	@Nonnull
+	final D stateDesc;
+
+	AbstractTtlStateVerifier(@Nonnull D stateDesc) {
+		this.stateDesc = stateDesc;
+	}
+
+	@Nonnull
+	static String randomString() {
+		StringBuilder sb = new StringBuilder();
+		IntStream.range(0, RANDOM.nextInt(14) + 2).forEach(i -> sb.append(randomChar()));
+		return sb.toString();
+	}
+
+	private static char randomChar() {
+		char d = (char) ('0' + RANDOM.nextInt(9));
+		char l = (char) ('a' + RANDOM.nextInt(25));
+		return RANDOM.nextBoolean() ? d : l;
+	}
+
+	@SuppressWarnings("unchecked")
+	@Override
+	@Nonnull
+	public State createState(@Nonnull FunctionInitializationContext context, @Nonnull StateTtlConfiguration ttlConfig) {
+		stateDesc.enableTimeToLive(ttlConfig);
+		return createState(context);
+	}
+
+	abstract State createState(FunctionInitializationContext context);
+
+	@SuppressWarnings("unchecked")
+	@Override
+	@Nonnull
+	public TypeSerializer<UV> getUpdateSerializer() {
+		return (TypeSerializer<UV>) stateDesc.getSerializer();
+	}
+
+	@SuppressWarnings("unchecked")
+	@Override
+	public GV get(@Nonnull State state) throws Exception {
+		return getInternal((S) state);
+	}
+
+	abstract GV getInternal(@Nonnull S state) throws Exception;
+
+	@SuppressWarnings("unchecked")
+	@Override
+	public void update(@Nonnull State state, Object update) throws Exception {
+		updateInternal((S) state, (UV) update);
+	}
+
+	abstract void updateInternal(@Nonnull S state, UV update) throws Exception;
+
+	@SuppressWarnings("unchecked")
+	@Override
+	public boolean verify(@Nonnull TtlVerificationContext<?, ?> verificationContextRaw, @Nonnull Time precision) {
```

Comment:
I think you can get rid of handling wildcards `<?, ?>` and casting by introducing the following method in `TtlVerifyFunction`:

```
	private <A, B> void verify(TtlStateVerifier<A, B> verifier, TtlVerificationContext<A, B> verificationContext) {
		verifier.verify(verificationContext, precision);
	}
```

and call it

```
verify(TtlStateVerifier.VERIFIERS_BY_NAME.get(value.getVerifierId()), new TtlVerificationContext<>(prevValues, value));
```